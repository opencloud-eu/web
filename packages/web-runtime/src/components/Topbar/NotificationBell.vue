<template>
  <oc-button
    id="oc-notifications-bell"
    v-oc-tooltip="notificationsLabel"
    class="relative"
    appearance="raw-inverse"
    color-role="chrome"
    :aria-label="notificationsLabel"
    no-hover
  >
    <oc-icon class="cursor-pointer flex items-center" name="notification-3" fill-type="line" />
    <span
      v-if="notificationCount"
      :key="notificationCount"
      :class="{ 'animate-shake transform-[translate3d(0, 0, 0)]': animate }"
      class="badge absolute top-[-6px] right-[-9px] p-1 text-xs leading-2 font-light text-center bg-red-600 text-white rounded-4xl box-content min-w-2 h-2 shadow-sm"
      v-text="notificationCountLabel"
    />
  </oc-button>
</template>
<script lang="ts">
import { computed } from 'vue'
import { useGettext } from 'vue3-gettext'
import { ref } from 'vue'
import { watch } from 'vue'

export default {
  props: {
    notificationCount: {
      type: Number,
      default: 0
    }
  },
  setup(props) {
    const { $gettext } = useGettext()
    const animate = ref(false)
    const notificationsLabel = computed(() => $gettext('Notifications'))
    const notificationCountLabel = computed(() => {
      if (props.notificationCount > 99) {
        return '99+'
      }
      return `${props.notificationCount}`
    })

    watch(
      () => props.notificationCount,
      () => {
        animate.value = true
        setTimeout(() => {
          animate.value = false
        }, 600)
      }
    )

    return {
      animate,
      notificationsLabel,
      notificationCountLabel
    }
  }
}
</script>
