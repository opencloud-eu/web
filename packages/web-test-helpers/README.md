# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
