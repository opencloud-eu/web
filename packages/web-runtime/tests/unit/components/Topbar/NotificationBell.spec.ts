import NotificationBell from '../../../../src/components/Topbar/NotificationBell.vue'
import { defaultComponentMocks, defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'

describe('NotificationBell', () => {
  it('should match snapshot', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should not render badge if 0 notifications', async () => {
    const { wrapper } = getWrapper({})
    await wrapper.setProps({ notificationCount: 0 })
    expect(wrapper.find('.badge').exists()).toBeFalsy()
  })
  it.each([
    [1, '1'],
    [11, '11'],
    [99, '99'],
    [110, '99+']
  ])("should render badge count '%s'", async (input, expected) => {
    const { wrapper } = getWrapper({})
    await wrapper.setProps({ notificationCount: input })
    expect(wrapper.find('.badge').text()).toBe(expected)
  })
  it('displays a tooltip with the notifications label', () => {
    const { wrapper } = getWrapper({
      mountType: mount
    })
    expect(wrapper.find('#oc-notifications-bell').attributes('aria-label')).toEqual('Notifications')
  })
  it('animates when notification count changes', async () => {
    const { wrapper } = getWrapper()
    wrapper.setProps({ notificationCount: 10 })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.shake').exists()).toBe(true)
  })
})
function getWrapper({ mountType = mount, mocks = {} } = {}) {
  const localMocks = { ...defaultComponentMocks(), ...mocks }

  return {
    mocks: localMocks,
    wrapper: mountType(NotificationBell, {
      global: {
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()],
        mocks: localMocks,
        stubs: { 'oc-icon': true }
      }
    })
  }
}
