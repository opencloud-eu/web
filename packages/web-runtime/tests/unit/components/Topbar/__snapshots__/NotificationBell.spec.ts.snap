// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationBell > should match snapshot 1`] = `
"<button data-v-25fb4be0="" type="button" aria-label="Notifications" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-raw-inverse" id="oc-notifications-bell">
  <!--v-if-->
  <!-- @slot Content of the button -->
  <oc-icon-stub data-v-25fb4be0="" class="oc-cursor-pointer oc-flex oc-flex-middle" name="notification-3" fill-type="line"></oc-icon-stub>
  <!--v-if-->
</button>"
`;
